File: 1172_가장 긴 증가하는 부분 수열/main.js
/*

2017 Dec 20, made by Moon

*/

var fs = require('fs');
var input = fs.readFileSync('/dev/stdin').toString().split('\n');
var n = parseInt(input[0]);
var c = input[1].split(' ');

function main(n, c){
	var ix, ixLen, dp = [];
  
    dp[0] = c[0];
    dp.min = c[0];
    dp.max = c[0];
  
    for(ix = 1, ixLen = n; ix < ixLen; ix++){
  	  if(c[ix] > dp.max){
          dp.push(c[ix]);
          dp.max = c[ix];
      }else if(c[ix] < dp.max && c[ix] > dp.min){
    	if(dp.indexOf(c[ix]) === -1){
      	  dp.push(c[ix]);
        }
      }else if(c[ix] < dp.min){
    	  dp.push(c[ix]);
          dp.min = c[ix];
      }else{}
    }
  
    console.log(dp.length);
}

main(n, c);

File: 1172_가장 긴 증가하는 부분 수열/main.c
/*

2017 Dec 20, made by Jeon

*/

#include <stdio.h>
#define MAX 1001

int main(void)
{
	int n, max = 0;
	int A[MAX] = { 0 };
	int D[MAX] = { 0 };

	scanf("%d", &n);

	for (int i = 1; i <= n; i++)
	{
		scanf("%d", &A[i]);
		D[i] = 1;
	}

	for (int i = 2; i <= n; i++)
	{
		int temp = 0;
		for (int x = i - 1; x >= 1; x--)
		{
			if (A[i] > A[x])
			{
				temp = D[x] + 1;
				if (D[i] < temp)
					D[i] = temp;
			}
		}
	}
	
	for (int i = 1; i <= n; i++)
		if (max < D[i])
			max = D[i];

	printf("%d\n", max);
	return 0;
}